query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala

```diff
@@ -395,75 +419,212 @@ final case class LabelCardinalityExec(queryContext: QueryContext,
-  *
-  * Additionally, MAX_ITEMSKETCH_MAP_SIZE bounds the number of elements the
-  *   sketch counts at any given time. If sketch.update(elt, count) is called
-  *   on a currently-uncounted element while the sketch is at full capacity, then
-  *   all counters are decreased by some median count-value, and any below zero
-  *   are removed to make room for the new element. Again: tread carefully when
-  *   metric cardinalities are similar.
-  *
-  * See here for more details:
-  *   https://datasketches.apache.org/docs/Frequency/FrequentItemsOverview.html
+  * Creates a map of (prefix -> cardinalities) according to the TsCardinalities LogicalPlan.
+  *   See TsCardinalities for more details.
   */
-final case class TopkCardExec(queryContext: QueryContext,
-                              dispatcher: PlanDispatcher,
-                              dataset: DatasetRef,
-                              shard: Int,
-                              shardKeyPrefix: Seq[String],
-                              k: Int,
-                              addInactive: Boolean) extends LeafExecPlan {
+final case class TsCardExec(queryContext: QueryContext,
+                            dispatcher: PlanDispatcher,
+                            dataset: DatasetRef,
+                            shard: Int,
+                            shardKeyPrefix: Seq[String],
+                            groupDepth: GroupDepth.Value) extends LeafExecPlan {
+  require(1 + groupDepth.prefixIndex >= shardKeyPrefix.size,
+    "groupDepth indicate a depth at least as deep as shardKeyPrefix")
 
   override def enforceLimit: Boolean = false
 
+  /**
+   * Given a shard key prefix, iterates through all child
+   *   prefixes with the specified size.
+   *
+   * Note: next() will always return the same array instance.
+   *
+   * @param tsMemStore contains the shard keys over which to iterate
+   * @param initPrefix shard key prefix to resolve
+   * @param extendToSize the length of the prefix children to iterate through.
+   *   Must be at (1) least initPrefix.size and (2) greater than zero.
+   */
+  class PrefixIterator (tsMemStore: TimeSeriesMemStore,
+                        initPrefix: Seq[String],
+                        extendToSize: Int) extends Iterator[Seq[String]]{
+    require(extendToSize > 0,
+      "extendToSize must be positive")
+    require(extendToSize >= initPrefix.size,
+      "extendToSize must be at least initPrefix.size")
+
+    // A queue for each set of "space" (i.e. namespace, workspace, metric) labels.
+    // Spaces are stored in descending precedence, and each successive queue
+    //   contains labels for the "subspace" of the label at the front of the
+    //   previous queue.
+    private val spaceQueues = new Array[mutable.Queue[String]](extendToSize)
+
+    // Stores the prefix returned by next().
+    private val currPrefix = new mutable.ArrayBuffer[String](extendToSize)
+
+    private var iFirstEmptyQueue = -1
+
+    initialize()
```

Comment:
This is like a pseudo-constructor, but it would seem gratuitous to initialize this in a companion object.